tools/symbols_map_generator.py

```diff
@@ -0,0 +1,278 @@
+#!/usr/bin/env python3
+
+# Copyright © Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 2 or 3
+# as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+import argparse
+from pathlib import Path
+from typing import Literal, TypedDict, get_args, Optional
+import clang.cindex
+
+_logger = logging.getLogger(__name__)
+
+LibraryName = Literal[
+    "miral", "mir_common", "mir_core", "miroil", "mir_platform",
+    "mir_renderer", "mir_renderers", "mir_server", "mir_test", "mir_wayland"]
+
+
+class LibraryInfo(TypedDict):
+    headers_dir: list[str]
+    map_file: str
+    include_headers: Optional[list[str]]
+
+
+# Directory paths are relative to the root of the Mir project
+library_to_data_map: dict[LibraryName, LibraryInfo] = {
+    "miral": {
+        "headers_dir": [
+            "include/miral"
+        ],
+        "map_file": "src/miral/symbols.map",
+    },
+    "mir_server": {
+        "headers_dir": [
+            "include/server/mir",
+            "src/include/server/mir"
+        ],
+        "map_file": "src/server/symbols.map",
+        "include_headers": [
+            "src/include/server"
+        ]
+    }
+}
```

Comment:
I feel like these should be command line parameters?
`--headers-dir` / `--internal-headers-dir` / `--symbol-map`?

Reply:
I also think that `--internal-headers-dir` is useless to be honest. Those are two different libraries being created by two different commands IMO

Reply:
How do we manage dependencies here actually? We need to tell clang which includes it should resolve. I'm not sure how to do this without hardcoded paths ATM

Reply:
> I also think that --internal-headers-dir is useless to be honest. Those are two different libraries being created by two different commands IMO

Wasn't the original point of this exercise to be able to spit out a `MIR_SERVER_INTERNAL` stanza for `libmirserver.so`? That'll be one library and one `symbols.map` file but with two disjoint sets of headers.

Reply:
> Wasn't the original point of this exercise to be able to spit out a `MIR_SERVER_INTERNAL` stanza

It was. Although _at present_ all the symbols from libmirserver should be `MIR_SERVER_INTERNAL`